src/java/fr/paris/lutece/plugins/forms/web/FormXPage.java

```diff
@@ -1129,20 +1129,21 @@ public XPage doSynchronousUploadDocument( HttpServletRequest request ) throws Si
      */
     private void saveFormResponse( Form form, HttpServletRequest request ) throws SiteMessageException
     {
-        /*
-         * The deletion of the synchronized block is due to the accumulation of http threads on the server when scaling up. The current implementation does not
-         * guarantee the following two checks in case of competitor access: -checkNumberMaxResponseForm (form, request); -checkIfUserResponseForm (form,
-         * request); This implementation to review and modify.
-         */
-        checkNumberMaxResponseForm( form, request );
-        checkIfUserResponseForm( form, request );
         FormResponse formResponse = _formResponseManager.getFormResponse( );
         if ( form.isAuthentificationNeeded( ) )
         {
+            checkIfUserResponseForm( form, request );
             LuteceUser user = SecurityService.getInstance( ).getRegisteredUser( request );
             formResponse.setGuid( user.getName( ) );
         }
-        _formService.saveForm( form, formResponse );
+        
+        synchronized ( FormXPage.class )
```

Comment:
This introduces a critical section around an expensive DB operation for all forms, wether or not they are configured to limit the number of responses.
Maybe a better way would be to lock the form in the DB via a `SELECT FOR UPDATE` statement, and only if the form requires it ?